File: v4/rust/iprange-livedb/src/publication/cleanup.rs
//! Exact discard of direct-publication artifacts before main publication.

use std::fs::File;

#[cfg(unix)]
use super::namespace::{regular_identity, regular_link_count, Directory, NamespaceError};
use super::namespace::{Identity, Name};
use super::output::{CreatedOutput, OutputAttempt, PreparedOutput};
use super::problem::Problem;
#[cfg(unix)]
use super::result::NameSlot;
use super::result::{ArtifactKind, CleanupArtifacts, Seed};
use super::{
    CleanupArtifact, DirectoryRole, Housekeeping, HousekeepingArtifact, PrivateOutputAttempt,
};

#[cfg(windows)]
#[path = "cleanup/windows.rs"]
mod windows;

#[derive(Clone, Copy)]
pub(super) enum ReservationLocation {
    Private,
    Canonical,
    Either,
}

#[derive(Clone, Copy)]
pub(super) struct ReservationOwner<'a> {
    pub(super) file: &'a File,
    pub(super) identity: Option<Identity>,
    pub(super) private_name: &'a Name,
    pub(super) location: ReservationLocation,
}

#[derive(Clone, Copy)]
pub(super) struct OutputOwner<'a> {
    pub(super) file: &'a File,
    pub(super) identity: Identity,
    pub(super) name: &'a Name,
}

pub(super) struct Summary {
    pub(super) artifacts: CleanupArtifacts,
    pub(super) housekeeping: Housekeeping,
    pub(super) visible_housekeeping: Vec<HousekeepingArtifact>,
    pub(super) main_absent: bool,
    pub(super) coordination_absent: bool,
}

pub(crate) struct EarlyDiscard {
    pub(crate) output: PrivateOutputAttempt,
    pub(crate) artifact: Option<CleanupArtifact>,
    pub(crate) housekeeping: Housekeeping,
    pub(crate) visible_housekeeping: Box<[HousekeepingArtifact]>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(super) enum Point {
    OutputRemoval,
    ReservationRemoval,
    #[cfg(unix)]
    DirectorySync,
}

pub(crate) fn discard_created(created: &CreatedOutput) -> EarlyDiscard {
    let facts = created.facts();
    #[cfg(windows)]
    {
        windows::discard_created(created, facts)
    }
    #[cfg(unix)]
    {
        let identity = facts
            .identity
            .and_then(|value| Identity::decode(value.bytes));
        let problem = discard_one(
            created.destination().directory(),
            created.name(),
            created.file(),
            identity,
        );
        let artifact = problem.map(|problem| early_artifact(&facts, problem));
        EarlyDiscard {
            output: facts,
            artifact,
            housekeeping: Housekeeping::None,
            visible_housekeeping: Box::default(),
        }
    }
}

pub(crate) fn discard_attempt(attempt: &OutputAttempt, file: &File) -> EarlyDiscard {
    let facts = attempt.facts();
    #[cfg(windows)]
    {
        windows::discard_attempt(attempt, file, facts)
    }
    #[cfg(unix)]
    {
        let problem = discard_one(
            attempt.destination().directory(),
            attempt.name(),
            file,
            Some(attempt.identity()),
        );
        let artifact = problem.map(|problem| early_artifact(&facts, problem));
        EarlyDiscard {
            output: facts,
            artifact,
            housekeeping: Housekeeping::None,
            visible_housekeeping: Box::default(),
        }
    }
}

pub(super) fn discard_with(
    seed: &mut Seed,
    output: &PreparedOutput,
    reservation: Option<ReservationOwner<'_>>,
    checkpoint: impl FnMut(Point) -> Result<(), Problem>,
) -> Summary {
    let destination = output.attempt.destination();
    discard_owners_with(
        seed,
        destination,
        Some(OutputOwner {
            file: &output.file,
            identity: output.attempt.identity(),
            name: output.attempt.name(),
        }),
        reservation,
        checkpoint,
    )
}

pub(super) fn discard_recovered(
    seed: &mut Seed,
    destination: &super::namespace::Destination,
    output: Option<OutputOwner<'_>>,
    reservation: Option<ReservationOwner<'_>>,
) -> Summary {
    discard_owners_with(seed, destination, output, reservation, |_| Ok(()))
}

fn discard_owners_with(
    seed: &mut Seed,
    destination: &super::namespace::Destination,
    output: Option<OutputOwner<'_>>,
    reservation: Option<ReservationOwner<'_>>,
    mut checkpoint: impl FnMut(Point) -> Result<(), Problem>,
) -> Summary {
    #[cfg(windows)]
    {
        windows::discard_owners(seed, destination, output, reservation, &mut checkpoint)
    }
    #[cfg(unix)]
    {
        let directory = destination.directory();
        let output_removal = output.map(|owner| {
            checkpoint(Point::OutputRemoval)
                .and_then(|()| remove_output(directory, owner))
                .unwrap_or_else(|problem| {
                    Removal::failed(
                        ArtifactKind::PrivateOutput,
                        NameSlot::PrivateOutput,
                        Some(owner.identity),
                        problem,
                    )
                })
        });
        let reservation_removal = reservation.map(|owner| {
            checkpoint(Point::ReservationRemoval)
                .and_then(|()| remove_reservation(directory, destination.coordination(), owner))
                .unwrap_or_else(|problem| {
                    Removal::failed(
                        ArtifactKind::PrivateReservation,
                        default_slot(owner.location),
                        owner.identity,
                        problem,
                    )
                })
        });
        let needs_sync = output_removal.as_ref().is_some_and(Removal::needs_sync)
            || reservation_removal
                .as_ref()
                .is_some_and(Removal::needs_sync);
        let sync = if needs_sync {
            checkpoint(Point::DirectorySync)
                .and_then(|()| directory.sync().map_err(|error| Problem::namespace(&error)))
                .and_then(|()| {
                    directory
                        .verify()
                        .map_err(|error| Problem::namespace(&error))
                })
                .err()
        } else {
            None
        };

        let mut artifacts = CleanupArtifacts::new();
        if let Some(removal) = output_removal {
            finish_removal(seed, removal, sync, &mut artifacts);
        }
        if let Some(removal) = reservation_removal {
            finish_removal(seed, removal, sync, &mut artifacts);
        }
        Summary {
            artifacts,
            housekeeping: Housekeeping::None,
            visible_housekeeping: Vec::new(),
            main_absent: directory.require_absent(destination.main()).is_ok(),
            coordination_absent: directory.require_absent(destination.coordination()).is_ok(),
        }
    }
}

#[cfg(unix)]
fn discard_one(
    directory: &Directory,
    name: &Name,
    file: &File,
    identity: Option<Identity>,
) -> Option<Problem> {
    let identity = match identity {
        Some(identity) => identity,
        None => {
            return Some(Problem::cleanup_conflict(
                "private output identity was not established",
            ))
        }
    };
    let removal = match remove_output(
        directory,
        OutputOwner {
            file,
            identity,
            name,
        },
    ) {
        Ok(removal) => removal,
        Err(problem) => return Some(problem),
    };
    finish_one(directory, removal)
}

#[cfg(unix)]
fn finish_one(directory: &Directory, removal: Removal<'_>) -> Option<Problem> {
    match removal.state {
        RemovalState::Failed(problem) => Some(problem),
        RemovalState::NeedsSync(file) => directory
            .sync()
            .and_then(|()| directory.verify())
            .map_err(|error| Problem::namespace(&error))
            .and_then(|()| match links(file)? {
                0 => Ok(()),
                _ => Err(Problem::cleanup_conflict(
                    "private output removal was not proved",
                )),
            })
            .err(),
    }
}

pub(super) fn early_artifact(facts: &PrivateOutputAttempt, error: Problem) -> CleanupArtifact {
    CleanupArtifact {
        kind: ArtifactKind::PrivateOutput,
        directory_role: DirectoryRole::Destination,
        directory_identity: facts.directory_identity,
        basename_encoding: facts.basename_encoding,
        basename: facts.basename.clone(),
        identity: facts.identity,
        creation_security: Some(facts.creation_security.clone()),
        unpublished_tail: None,
        error,
    }
}

#[cfg(unix)]
fn remove_output<'a>(
    directory: &Directory,
    output: OutputOwner<'a>,
) -> Result<Removal<'a>, Problem> {
    remove(
        directory,
        output.file,
        output.identity,
        ArtifactKind::PrivateOutput,
        [Some((output.name, NameSlot::PrivateOutput)), None],
    )
}

#[cfg(unix)]
fn remove_reservation<'a>(
    directory: &Directory,
    canonical: &'a Name,
    owner: ReservationOwner<'a>,
) -> Result<Removal<'a>, Problem> {
    let identity = match owner.identity {
        Some(identity) => identity,
        None => regular_identity(owner.file, directory.identity())
            .map_err(|error| Problem::namespace(&error))?,
    };
    let names = match owner.location {
        ReservationLocation::Private => [
            Some((owner.private_name, NameSlot::PrivateReservation)),
            None,
        ],
        ReservationLocation::Canonical => [
            Some((canonical, NameSlot::Coordination)),
            Some((owner.private_name, NameSlot::PrivateReservation)),
        ],
        ReservationLocation::Either => [
            Some((owner.private_name, NameSlot::PrivateReservation)),
            Some((canonical, NameSlot::Coordination)),
        ],
    };
    remove(
        directory,
        owner.file,
        identity,
        ArtifactKind::PrivateReservation,
        names,
    )
}

#[cfg(unix)]
fn remove<'a>(
    directory: &Directory,
    file: &'a File,
    identity: Identity,
    kind: ArtifactKind,
    names: [Option<(&'a Name, NameSlot)>; 2],
) -> Result<Removal<'a>, Problem> {
    let default_name = names[0].expect("cleanup requires one exact name").1;
    match links(file)? {
        0 => return Ok(Removal::awaiting_sync(kind, default_name, identity, file)),
        1 => {}
        _ => {
            return Err(Problem::cleanup_conflict(
                "owned publication artifact has unexpected links",
            ))
        }
    }

    match unlink_names(directory, identity, names)? {
        Some(slot) => require_unlinked(kind, slot, identity, file),
        None if links(file)? == 0 => Ok(Removal::awaiting_sync(kind, default_name, identity, file)),
        None => Err(Problem::cleanup_conflict(
            "owned publication artifact has no exact retained name",
        )),
    }
}

#[cfg(unix)]
fn unlink_names(
    directory: &Directory,
    identity: Identity,
    names: [Option<(&Name, NameSlot)>; 2],
) -> Result<Option<NameSlot>, Problem> {
    let mut first_problem = None;
    for (name, slot) in names.into_iter().flatten() {
        match directory.unlink_exact(name, identity) {
            Ok(true) => return Ok(Some(slot)),
            Ok(false)
            | Err(NamespaceError::Missing | NamespaceError::IdentityChanged)
            | Err(NamespaceError::NotRegular) => {}
            Err(error @ NamespaceError::LinkCount(_)) => return Err(Problem::namespace(&error)),
            Err(error) => {
                first_problem.get_or_insert_with(|| Problem::namespace(&error));
            }
        }
    }
    match first_problem {
        Some(problem) => Err(problem),
        None => Ok(None),
    }
}

#[cfg(unix)]
fn require_unlinked<'a>(
    kind: ArtifactKind,
    name: NameSlot,
    identity: Identity,
    file: &'a File,
) -> Result<Removal<'a>, Problem> {
    if links(file)? != 0 {
        return Err(Problem::cleanup_conflict(
            "unlinked publication artifact still has links",
        ));
    }
    Ok(Removal::awaiting_sync(kind, name, identity, file))
}

#[cfg(unix)]
fn links(file: &File) -> Result<u64, Problem> {
    regular_link_count(file).map_err(|error| Problem::namespace(&error))
}

#[cfg(unix)]
fn finish_removal(
    seed: &mut Seed,
    removal: Removal<'_>,
    sync: Option<Problem>,
    artifacts: &mut CleanupArtifacts,
) {
    let problem = match removal.state {
        RemovalState::Failed(problem) => Some(problem),
        RemovalState::NeedsSync(file) => sync.or_else(|| match links(file) {
            Ok(0) => None,
            Ok(_) => Some(Problem::cleanup_conflict(
                "publication artifact removal was not proved",
            )),
            Err(problem) => Some(problem),
        }),
    };
    if let Some(problem) = problem {
        artifacts.push(seed.artifact(removal.kind, removal.name, removal.identity, problem));
    }
}

#[cfg(unix)]
struct Removal<'a> {
    kind: ArtifactKind,
    name: NameSlot,
    identity: Option<Identity>,
    state: RemovalState<'a>,
}

#[cfg(unix)]
impl<'a> Removal<'a> {
    fn awaiting_sync(
        kind: ArtifactKind,
        name: NameSlot,
        identity: Identity,
        file: &'a File,
    ) -> Self {
        Self {
            kind,
            name,
            identity: Some(identity),
            state: RemovalState::NeedsSync(file),
        }
    }

    fn failed(
        kind: ArtifactKind,
        name: NameSlot,
        identity: Option<Identity>,
        problem: Problem,
    ) -> Self {
        Self {
            kind,
            name,
            identity,
            state: RemovalState::Failed(problem),
        }
    }

    fn needs_sync(&self) -> bool {
        matches!(self.state, RemovalState::NeedsSync(_))
    }
}

#[cfg(unix)]
enum RemovalState<'a> {
    NeedsSync(&'a File),
    Failed(Problem),
}

#[cfg(unix)]
const fn default_slot(location: ReservationLocation) -> NameSlot {
    match location {
        ReservationLocation::Private | ReservationLocation::Either => NameSlot::PrivateReservation,
        ReservationLocation::Canonical => NameSlot::Coordination,
    }
}

File: v4/go/internal/publication/cleanup.go
// Exact discard of direct-publication artifacts before main
// publication (Rust publication/cleanup.rs). One removal proves the
// retained name went away, then the directory sync and verify run; a
// removal that cannot be proved pushes one exact cleanup artifact into
// the result ledger. The posix arm runs the unlink machine; the
// windows arm runs the authenticated GC transition
// (cleanup_discard_windows.go).

package publication

import (
	"os"

	"github.com/firehol/iprange/v4/go/internal/live"
)

// identityOptional is the Go peer of Rust Option<Identity> for the
// cleanup machine: a value presence flag that keeps every success
// path on the stack (Rust Copy semantics).
type identityOptional struct {
	identity live.FileIdentity
	present  bool
}

// ownerLocation classifies the name set one cleanup reservation owner
// may hold (Rust cleanup.rs ReservationLocation: Private, Canonical,
// or Either). The reservation custody position of reservation_verify.go
// keeps the reservationLocation name.
type ownerLocation uint8

const (
	ownerLocationPrivate ownerLocation = iota
	ownerLocationCanonical
	ownerLocationEither
)

// reservationOwner is one still-owned private reservation (Rust
// ReservationOwner). identity is absent when the owner cannot present
// the reservation identity yet (Rust Option<Identity>).
type reservationOwner struct {
	file        *os.File
	identity    identityOptional
	privateName string
	location    ownerLocation
}

// outputOwner is one still-owned private output (Rust OutputOwner).
type outputOwner struct {
	file     *os.File
	identity live.FileIdentity
	name     string
}

// cleanupSummary is the factual outcome of one discard (Rust
// Summary): the cleanup ledger, the absent-main and absent-
// coordination proofs, and the housekeeping facts (always none on
// unix).
type cleanupSummary struct {
	artifacts           CleanupArtifacts
	housekeeping        Housekeeping
	visibleHousekeeping []HousekeepingArtifact
	mainAbsent          bool
	coordinationAbsent  bool
}

// earlyDiscard is the factual outcome of one pre-publication discard
// of a created or attempted output (Rust EarlyDiscard).
type earlyDiscard struct {
	output              PrivateOutputAttempt
	artifact            *CleanupArtifact
	housekeeping        Housekeeping
	visibleHousekeeping []HousekeepingArtifact
}

// cleanupPoint is one discard checkpoint (Rust Point).
type cleanupPoint uint8

const (
	cleanupPointOutputRemoval cleanupPoint = iota
	cleanupPointReservationRemoval
	cleanupPointDirectorySync
)

// discardCreated discards one freshly created output whose identity
// is best-effort (Rust discard_created).

// discardAttempt discards one secured output attempt whose identity
// is established (Rust discard_attempt).

// failedAttempt records one already-failed attempt (Rust
// failed_attempt): the problem becomes the exact artifact with no
// further namespace work.
func failedAttempt(facts PrivateOutputAttempt, problem error) earlyDiscard {
	a := earlyArtifact(&facts, problem)
	return earlyDiscard{output: facts, artifact: &a, housekeeping: HousekeepingNone}
}

// confirmedAbsent records one attempt proved absent (Rust
// confirmed_absent): no artifact, no namespace work.
func confirmedAbsent(facts PrivateOutputAttempt) earlyDiscard {
	return earlyDiscard{output: facts, housekeeping: HousekeepingNone}
}

// discardCreated discards one freshly created output whose identity
// is best-effort (Rust discard_created; the platform arm selects the
// unix removal or the Windows GC transition).
func discardCreated(created *createdOutput) earlyDiscard {
	return discardCreatedPlatform(created)
}

// discardAttempt discards one secured output attempt whose identity
// is established (Rust discard_attempt; platform arm).
func discardAttempt(attempt *outputAttempt, file *os.File) earlyDiscard {
	return discardAttemptPlatform(attempt, file)
}

// discardOwnersWith is the shared discard core (Rust
// discard_owners_with; the platform arm selects the unix removal
// machine or the Windows GC transition).
func discardOwnersWith(s *seed, destination *destination, output outputOwner, hasOutput bool, reservation reservationOwner, hasReservation bool, checkpoint func(cleanupPoint) error) cleanupSummary {
	return discardOwnersWithPlatform(s, destination, output, hasOutput, reservation, hasReservation, checkpoint)
}

// discardWith discards one prepared output and its optional
// reservation (Rust discard_with): every removal runs behind its
// checkpoint.
func discardWith(s *seed, output *preparedOutput, reservation *reservationOwner, checkpoint func(cleanupPoint) error) cleanupSummary {
	d := output.attempt.destinationOf()
	identity := output.attempt.identityOf()
	owner := outputOwner{file: output.file, identity: identity, name: output.attempt.nameOf()}
	var res reservationOwner
	hasReservation := reservation != nil
	if hasReservation {
		res = *reservation
	}
	return discardOwnersWith(s, d, owner, true, res, hasReservation, checkpoint)
}

// discardRecovered discards one recovered output and reservation with
// no checkpoint (Rust discard_recovered passes the always-ok closure,
// so recovery paths never re-sync through this machine).
func discardRecovered(s *seed, destination *destination, output *outputOwner, reservation *reservationOwner) cleanupSummary {
	var owner outputOwner
	hasOutput := output != nil
	if hasOutput {
		owner = *output
	}
	var res reservationOwner
	hasReservation := reservation != nil
	if hasReservation {
		res = *reservation
	}
	return discardOwnersWith(s, destination, owner, hasOutput, res, hasReservation, func(cleanupPoint) error { return nil })
}

// discardOwnersWith is the shared discard core (Rust
// discard_owners_with): each owner runs behind its checkpoint, one
// directory sync proves every successful removal, and each failing
// removal pushes one exact artifact into the ledger.

// discardOne proves one private output fully gone (Rust discard_one):
// the identity must be established, then the output removal runs and
// the directory sync proves it.

// finishOne finishes one early removal by syncing the directory and
// proving zero remaining links (Rust finish_one). The sync and verify
// failures fold to namespace problems; the link proof failure is the
// fixed cleanup conflict.

// earlyArtifact builds the exact artifact of one early discard (Rust
// early_artifact): always the private output of the destination
// directory, carrying the portable attempt facts and the fixed
// problem.
func earlyArtifact(facts *PrivateOutputAttempt, problem error) CleanupArtifact {
	var identity *LocalFileIdentity
	if facts.IdentityPresent {
		converted := facts.Identity
		identity = &converted
	}
	security := facts.CreationSecurity
	return CleanupArtifact{
		Kind:              ArtifactPrivateOutput,
		DirectoryRole:     DirectoryRoleDestination,
		DirectoryIdentity: facts.DirectoryIdentity,
		BasenameEncoding:  facts.BasenameEncoding,
		Basename:          append([]byte(nil), facts.Basename...),
		Identity:          identity,
		CreationSecurity:  &security,
		UnpublishedTail:   nil,
		Error:             problem,
	}
}

// removeOutput builds the removal of one private output (Rust
// remove_output): the private output name is the single candidate.

// removeReservation builds the removal of one private reservation
// (Rust remove_reservation): the candidate name set follows the
// owner's location; a missing owner identity is re-proved from the
// open file.

// nameEntry is one exact removal candidate (Rust Option<(&Name,
// NameSlot)>): the name to unlink and the slot its success proves.
type nameEntry struct {
	name string
	slot nameSlot
	ok   bool
}

// removeFile runs one exact removal (Rust remove): zero links need no
// unlink, one link unlinks the candidate names in order, and any
// other link count is a fixed cleanup conflict.

// unlinkNames unlinks the first candidate name that still names the
// expected identity (Rust unlink_names). Absent, replaced, and non-
// regular candidates are skipped; a changed link count is a hard
// namespace error; any other failure is kept as the first problem and
// reported only when no name could be unlinked.

// requireUnlinked proves the inode has zero links after the exact
// unlink (Rust require_unlinked).

// links reports the current link count of one open file (Rust links);
// failures fold to namespace problems.

// finishRemoval closes one removal into the ledger (Rust
// finish_removal): a failing removal pushes one exact artifact that
// consumes its name slot from the seed.
func finishRemoval(s *seed, r removal, syncProblem error, artifacts *CleanupArtifacts) {
	var problem error
	switch {
	case r.state.problem != nil:
		problem = r.state.problem
	case syncProblem != nil:
		problem = syncProblem
	default:
		count, err := links(r.state.file)
		switch {
		case err != nil:
			problem = err
		case count != 0:
			problem = cleanupConflictProblem("publication artifact removal was not proved")
		}
	}
	if problem != nil {
		artifacts.Push(s.artifact(r.kind, r.name, r.identity, problem))
	}
}

// links reports the current link count of one open file (Rust links);
// failures fold to namespace problems. The shared finishRemoval uses
// it on every platform.
func links(file *os.File) (uint64, error) {
	count, err := live.RegularLinkCount(file)
	if err != nil {
		return 0, namespaceProblem(err)
	}
	return count, nil
}

// removal is one exact removal decision of the discard machine (Rust
// Removal): the artifact kind, the proved name slot, the removal
// identity, and the state (needs directory sync, or already failed).
type removal struct {
	kind     ArtifactKind
	name     nameSlot
	identity identityOptional
	state    removalState
}

// removalState is the state of one removal (Rust RemovalState): at
// most one of file and problem is set.
type removalState struct {
	file    *os.File // non-nil: needs the directory sync to prove removal
	problem error    // non-nil: the removal already failed
}

// awaitingSyncRemoval marks one removal that only needs the directory
// sync to prove the name went away (Rust Removal::awaiting_sync).
func awaitingSyncRemoval(kind ArtifactKind, slot nameSlot, identity live.FileIdentity, file *os.File) removal {
	return removal{
		kind:     kind,
		name:     slot,
		identity: identityOptional{present: true, identity: identity},
		state:    removalState{file: file},
	}
}

// failedRemoval marks one removal that already failed (Rust
// Removal::failed). identity is absent when the removal never
// established the inode identity.
func failedRemoval(kind ArtifactKind, slot nameSlot, identity identityOptional, problem error) removal {
	return removal{kind: kind, name: slot, identity: identity, state: removalState{problem: problem}}
}

// needsSync reports whether the removal still needs the directory
// sync (Rust Removal::needs_sync).
func (r removal) needsSync() bool { return r.state.file != nil }

// defaultSlot reports the artifact name slot of one reservation
// location (Rust default_slot).
func defaultSlot(location ownerLocation) nameSlot {
	switch location {
	case ownerLocationPrivate, ownerLocationEither:
		return nameSlotPrivateReservation
	default:
		return nameSlotCoordination
	}
}
